ludwig/data/split.py

```diff
@@ -0,0 +1,222 @@
+#! /usr/bin/env python
+# Copyright (c) 2022 Predibase, Inc.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import logging
+import warnings
+from abc import ABC, abstractmethod
+from typing import Any, Dict, List, Optional, Tuple
+
+import numpy as np
+
+from ludwig.backend.base import Backend
+from ludwig.constants import BINARY, CATEGORY, COLUMN, DATE, SPLIT, TYPE
+from ludwig.utils.data_utils import split_dataset_ttv
+from ludwig.utils.misc_utils import set_random_seed
+from ludwig.utils.registry import Registry
+from ludwig.utils.types import DataFrame, Series
+
+split_registry = Registry()
+default_random_seed = 42
+
+
+TMP_SPLIT_COL = "__SPLIT__"
+DEFAULT_PROBABILITIES = (0.7, 0.1, 0.2)
+
+
+class Splitter(ABC):
+    @abstractmethod
+    def split(
+        self, df: DataFrame, backend: Backend, random_seed: float = default_random_seed
+    ) -> Tuple[DataFrame, DataFrame, DataFrame]:
+        pass
+
+    def validate(self, config: Dict[str, Any]):
+        pass
+
+    def has_split(self, split_index: int) -> bool:
+        return True
+
+    @property
+    def required_columns(self) -> List[str]:
+        return []
+
+
+@split_registry.register("random", default=True)
+class RandomSplitter(Splitter):
+    def __init__(self, probabilities: List[float] = DEFAULT_PROBABILITIES, **kwargs):
+        self.probabilities = probabilities
+
+    def split(
+        self, df: DataFrame, backend: Backend, random_seed: float = default_random_seed
+    ) -> Tuple[DataFrame, DataFrame, DataFrame]:
+        if backend.df_engine.partitioned:
+            # The below approach is very inefficient for partitioned backends, which
+            # can split by partition. This may not be exact in all cases, but is much more efficient.
+            return df.random_split(self.probabilities, random_state=random_seed)
+
+        n = len(df)
+        d1 = int(self.probabilities[0] * n)
+        d2 = d1 + int(self.probabilities[1] * n)
+        return np.split(df.sample(frac=1, random_state=random_seed), [d1, d2])
+
+    def has_split(self, split_index: int) -> bool:
+        return self.probabilities[split_index] > 0
+
+
+@split_registry.register("fixed")
+class FixedSplitter(Splitter):
+    def __init__(self, column: str = SPLIT, **kwargs):
```

Comment:
At some point (perhaps in separate PR), what do you think about extending this to enable users to specify custom values for fixed splits?

For example, I can imagine users with an explicit split column, but the values in the column are like "TRAIN", "TEST" instead of 0, 1, and 2.

(default)
```yaml
split:
  type: fixed
  column: custom_split_column
  train_values: [0]
  validation_values: [1]
  test_values: [2]
```

(custom)
```yaml
split:
  type: fixed
  column: custom_split_column
  train_values: [train]
  validation_values: [test]
  test_values: [heldout]
```

Related: #1823

Reply:
Yes, I think we should definitely do this. Can be in a follow-up PR right after.